File: core/src/main/scala/org/onemodel/model/Attribute.scala
/*  This file is part of OneModel, a program to manage knowledge.
    Copyright in each year of 2004, 2010, 2011, and 2013-2016 inclusive, Luke A. Call; all rights reserved.
    OneModel is free software, distributed under a license that includes honesty, the Golden Rule, guidelines around binary
    distribution, and the GNU Affero General Public License as published by the Free Software Foundation, either version 3
    of the License, or (at your option) any later version.  See the file LICENSE for details.
    OneModel is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License along with OneModel.  If not, see <http://www.gnu.org/licenses/>

  ---------------------------------------------------
  If we ever do port to another database, create the Database interface (removed around 2014-1-1 give or take) and see other changes at that time.
  An alternative method is to use jdbc escapes (but this actually might be even more work?):  http://jdbc.postgresql.org/documentation/head/escapes.html  .
  Another alternative is a layer like JPA, ibatis, hibernate  etc etc.

*/
package org.onemodel.model

import org.onemodel.database.PostgreSQLDatabase


object Attribute {
  // unlike in Controller, these are intentionally a little different, for displaying also the day of the week:
  val DATEFORMAT = new java.text.SimpleDateFormat("EEE yyyy-MM-dd HH:mm:ss:SSS zzz")
  val DATEFORMAT_WITH_ERA = new java.text.SimpleDateFormat("EEE GGyyyy-MM-dd HH:mm:ss:SSS zzz")

  def usefulDateFormat(d: Long): String = {
    // No need to print "AD" unless we're really close?, as in this example:
    //scala > val DATEFORMAT_WITH_ERA = new java.text.SimpleDateFormat("GGyyyy-MM-dd HH:mm:ss:SSS zzz")
    //scala > DATEFORMAT_WITH_ERA.parse("ad 1-03-01 00:00:00:000 GMT").getTime //i.e., Jan 3, 1 AD.
    //res100: Long = -62130672000000
    if (d > -62130672000000L) DATEFORMAT.format(d)
    else DATEFORMAT_WITH_ERA.format(d)
  }

  def limitDescriptionLength(input: String, lengthLimitIn: Int): String = {
    if (lengthLimitIn != 0 && input.length > lengthLimitIn) {
      input.substring(0, lengthLimitIn - 3) + "..."
    } else input
  }

}
/**
 * Represents one attribute object in the system (usually [always, as of 1/2004] used as an attribute on a Entity).
 * Originally created as a place to put common stuff between Relation/QuantityAttribute/TextAttribute.
 */
abstract class Attribute(mDB: PostgreSQLDatabase, mId: Long) {
  // idea: somehow use scala features better to make it cleaner, so we don't need these extra 2 vars, because they are
  // used in 1-2 instances, and ignored in the rest.  One thing is that RelationToEntity and RelationToGroup are Attributes. Should they be?
  def getDisplayString(inLengthLimit: Int, parentEntity: Option[Entity], inRTId: Option[RelationType], simplify: Boolean = false): String

  protected def readDataFromDB()

  def delete()

  private[onemodel] def getIdWrapper: IdWrapper = {
    new IdWrapper(mId)
  }

  def getId: Long = {
    mId
  }

  def getFormId: Int = {
    PostgreSQLDatabase.getAttributeFormId(this.getClass.getSimpleName)
  }

  protected def assignCommonVars(parentIdIn: Long, attrTypeIdIn: Long, sortingIndexIn: Long) {
    mParentId = parentIdIn
    mAttrTypeId = attrTypeIdIn
    mSortingIndex = sortingIndexIn
    mAlreadyReadData = true
  }

  def getAttrTypeId: Long = {
    if (!mAlreadyReadData) readDataFromDB()
    mAttrTypeId
  }

  def getSortingIndex: Long = {
    if (!mAlreadyReadData) readDataFromDB()
    mSortingIndex
  }

  // idea: make the scope definitions (by whatever name: "private[onemodel] ") sensible and uniform
  private[onemodel] def getParentId: Long = {
    if (!mAlreadyReadData) readDataFromDB()
    mParentId
  }

  /**
   * For descriptions of the meanings of these variables, see the comments
   * on PostgreSQLDatabase.createTables(...), and examples in the database testing code.
   */
  protected var mParentId: Long = 0L
  protected var mAttrTypeId: Long = 0L
  protected var mAlreadyReadData: Boolean = false
  protected var mSortingIndex: Long = 0L
}

File: core/src/model/attribute.rs
/* . This file is part of OneModel, a program to manage knowledge.
    Copyright in each year of 2004, 2010, 2011, 2013-2017 inclusive, and 2023, Luke A Call.
    OneModel is free software, distributed under a license that includes honesty, the Golden Rule,
    and the GNU Affero General Public License as published by the Free Software Foundation;
    see the file LICENSE for license version and details.
    OneModel is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License along with OneModel.  If not, see <http://www.gnu.org/licenses/>
*/
// use anyhow::{anyhow, Result};
// use chrono::LocalResult;
// use chrono::prelude::*;
// use crate::util::Util;
use crate::model::entity::Entity;
// use crate::model::id_wrapper::IdWrapper;
use crate::model::relation_type::RelationType;
use sqlx::{Postgres, Transaction};
use std::cell::{RefCell};
use std::rc::Rc;

/// Represents one attribute object in the system (usually [always, as of 1/2004] used as an attribute on a Entity).
/// Originally created as a place to put common stuff between Relation/QuantityAttribute/TextAttribute.
pub trait Attribute {
    //%%?:
    // Idea: somehow use language features better to make it cleaner, so we don't need these extra 2 vars, because they are
    // used in 1-2 instances, and ignored in the rest.  One thing is that RelationTo[Local|Remote]Entity and RelationToGroup
    // are Attributes. Should they be?

    fn get_display_string(
        &mut self,
        in_length_limit: usize,
        parent_entity: Option<Entity>,
        in_rt_id: Option<RelationType>,
        simplify: bool, /* = false*/
    ) -> Result<String, anyhow::Error>;

    fn read_data_from_db(
        &mut self,
        transaction: Option<Rc<RefCell<Transaction<Postgres>>>>,
    ) -> Result<(), anyhow::Error>;

    fn delete<'a>(
        &'a self,
        transaction: Option<Rc<RefCell<Transaction<'a, Postgres>>>>,
    ) -> Result<u64, anyhow::Error>;

    //looks unused except for Entity and EntityClass
    // fn get_id_wrapper(&self) -> IdWrapper;

    fn get_id(&self) -> i64;

    fn get_form_id(&self) -> Result<i32, anyhow::Error>;

    // fn assign_common_vars(parent_id_in: i64, attr_type_id_in: i64, sorting_index_in: i64);

    fn get_attr_type_id(
        &mut self,
        transaction: Option<Rc<RefCell<Transaction<Postgres>>>>,
    ) -> Result<i64, anyhow::Error>;

    fn get_sorting_index(
        &mut self,
        transaction: Option<Rc<RefCell<Transaction<Postgres>>>>,
    ) -> Result<i64, anyhow::Error>;

    fn get_parent_id(
        &mut self,
        transaction: Option<Rc<RefCell<Transaction<Postgres>>>>,
    ) -> Result<i64, anyhow::Error>;


    // For descriptions of the meanings of these variables, see the comments
    // on create_tables(...), and examples in the database testing code &/or in PostgreSQLDatabase or Database classes.
    // %%put these in the structs implementing this trait, along w/ those above methods!
    //db: Database;
    //id: i64;
    // protected let mut parent_id: i64 = 0L;
    // protected let mut attr_type_id: i64 = 0L;
    // protected let mut already_read_data: bool = false;
    // protected let mut sorting_index: i64 = 0L;
}
